### ISSUE TITLE
[FR] Please allow to use defines as benchmark names

### ISSUE THREAD

**Issue by emmenlau** [google/benchmark#1053]
**Is your feature request related to a problem? Please describe.**
We have a number of benchmarks that compare different implementations of an algorithm that has a standard interface. Therefore the benchmark code between these implementations is identical. Instead of duplicating the benchmark, we use defines to select the implementation and then include the actual benchmark code. This works very well, with one problem: The name of a benchmark in the macro can not itself be a define. So the following does not work:
```
#define MY_BECHMARK_NAME SomeBenchmark
BENCHMARK_DEFINE_F(MY_BECHMARK_NAME, TestSomething)(benchmark::State& aState) {
    // code goes here...
}
```

**Describe the solution you'd like**
I'm under the impression that the above example code should and could work. The problem seems to come from the macro stringification in for example `BENCHMARK_DEFINE_F`, where a separate macro invocation would be required to stringify the names. See for example the bottom example in https://gcc.gnu.org/onlinedocs/gcc-4.8.5/cpp/Stringification.html

**Describe alternatives you've considered**
Instead of using a define for the benchmark name, we could use functions that are called. These functions can be shared between implementations. However the actual benchmark macros will need to be repeated, and they constitute already a significant number of lines. We want to avoid diverging implementations, and maintaining all the macro invocations for two or three otherwise identical benchmarks becomes tedious and error prone.

**Comment by pleroy** [comment]
It seems to me that:
```
using MyBechmarkName = SomeBenchmark;
BENCHMARK_DEFINE_F(MyBechmarkName, TestSomething)(benchmark::State& aState) {...}
```
would do what you want without messing with macros.

**Comment by emmenlau** [comment]
Dear @pleroy thanks for the very quick response! I tried your suggestion but I'm not sure that works. I get error:
```
SomeBenchmark.cc:26:62: error: ‘SomeBenchmark’ does not name a type
 using MyBechmarkName = SomeBenchmark;
                        ^~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by emmenlau** [comment]
Ok then I did it wrong! Now I tried with a fixture class name. But then it seems the macro binding is stronger than the using-statement? I get error:
```
SomeBenchmark.cc:44:20: error: ‘MyBechmarkName_TestSomething_Benchmark’ has not been declared
 BENCHMARK_DEFINE_F(MyBechmarkName, TestSomething)(benchmark::State& aState) {
                    ^
```

**Comment by pleroy** [comment]
It's hard to say without more context.  The class `MyBechmarkName_TestSomething_Benchmark` should be declared by `BENCHMARK_DEFINE_F`.  Maybe looking at the macro-expanded source would shed some light on the problem.

**Comment by emmenlau** [comment]
Hmm, after giving it more thought, how could the solution with `using` even work? The macros are expanded from the preprocessor, but the using-statement is handled by the processor, isn't it? If this is correct (which I assume) then the macro will always supersede the using-statement and the name will long be concatenated before the using is even parsed.

**Comment by pleroy** [comment]
So?  The preprocessor will generate something like:
```
class MyBechmarkName_TestSomething_Benchmark : public MyBechmarkName {...}
```
and the compiler will then resolve `MyBechmarkName` through the type alias.

Googletest [works exactly like that](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#death-test-naming).  If the Benchmark fixtures don't support this use case, then *that* is a bug (I believe they do, though).

**Comment by emmenlau** [comment]
Ok maybe I'm confused, but I don't see the picture yet. I'm under the impression that an offensive definition is for example here:
https://github.com/google/benchmark/blob/ffe1342eb2faa7d2e7c35b4db2ccf99fab81ec20/include/benchmark/benchmark.h#L1227-L1229

The name `void BaseClass##_##Method##_Benchmark::BenchmarkCase` will be expanded already by the preprocessor. Doesn't this already break the idea with the `using` statement?

**Comment by dominichamon** [comment]
closed via #1054

**Comment by emmenlau** [comment]
Thanks @dominichamon !
